### ISSUE TITLE
proxy tile improvements

### ISSUE THREAD

**Issue by simahawk** [plone/plone.app.mosaic#57]
* move it to transient tile
* rename it to "Content proxy"
* can't we use UID for the relation? What's the reason for using IntId?
* as for "content listing" we should be able to select a view for rendering it
* make possible to provide a custom selector and not rely always on "#content" element

**Comment by djay** [comment]
I don't think the word proxy is likely makes much sense to the user at all.
Maybe "reuse content"?

**Comment by datakurre** [comment]
@simahawk

Rename +1

UID +1 (the initial implementation is just so old that it uses intids)

Maybe we should restore IStandardtilesSettings registry for being able to define those available renderings TTW. (I think, it was removed by SeaSprint. Yet, it's a bit complex that we have separate registry settings for p.a.blocks, p.a.tiles, p.a.standardtiles and p.a.mosaic.)

An alternative for #content would be to rely on content-core macro (or allow to define the used macro).

**Comment by datakurre** [comment]
Dylan Jay wrote:
> I don't think the word proxy is likely makes much sense to the user at all.
> Maybe "reuse content"?

The "tile" is currently under "Insert"-menu, so the name should be noun.

Maybe just "Content" then, or "Existing content".

**Comment by djay** [comment]
"Existing content" is good I think.

**Comment by simahawk** [comment]
I'm working on the branch `simahawk-proxy-imp` but I'm getting an error on autocomplete of the `content_uid` field. Looks like the pattern widget is fails when generating search results vocabulary:

```
2015-05-03 12:33:47 ERROR Zope.SiteErrorLog 1430649227.260.660689242176 http://localhost:8080/Plone/folder-1/lorem-ipsum-dolor-sit-amet-inani-voluptatum-per-in-ei-ius-esse-vulputate-sit-eu-nonumy-ornatus/@@add-tile/plone.app.standardtiles.existingcontent/++widget++plone.app.standardtiles.existingcontent.content_uid/@@getSource
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module plone.app.content.browser.vocabulary, line 151, in __call__
TypeError: 'NoneType' object has no attribute '__getitem__'
```
Looking at code I think it simply doesn't expect to hava an UID as value.
The field is defined here https://github.com/plone/plone.app.standardtiles/blob/simahawk-proxy-imp/plone/app/standardtiles/existingcontent.py#L16

Here I'm using contenttree widget and is working fine (different field type though) https://github.com/simahawk/collective.compositepage/blob/imp-and-fixes/collective/compositepage/tiles/related.py#L24

**Comment by datakurre** [comment]
@simahawk 

TLDR; Just use Choice instead of RelationChoice. It's the only option for a transient tile.

RelationChoice excepts objects and saves them as Intid based RelationValues. Of course, ideally those would allow to track all uses of content objects, create backlinks, detect possible breakages when deleting objects, but it I recall it working properly only when the context can be acquired properly. So, at first, RelationChoice would only work with persistent tiles and even then it would not work perfectly, because RelationValues would be saved into dictionary, which does not implement ILocation (does not know parent, or is never acquisition wrapper).

Personally, I'd like to replace zc.relations with something simpler. David Glick once said he had almost working implementation of UUID based RelationIndex for ZCatalog / PortalCatalog, but... that said, p.a.contenttypes' related items behavior uses zc.relations, so it will be with us for some time.

**Comment by simahawk** [comment]
@datakurre ok tnx, then I need to check how to enable the same widget (related items) onto it :)

**Comment by simahawk** [comment]
using schema.Choice but still getting the same error from `plone.app.content.browser.vocabulary`.
The problem is that it expects brains. Is there a way for passing parameters for making it behave properly? Or, should we use another source? BTW: thing of sources in `plone.formwidget.contenttree` ... is really a mess :S

**Comment by datakurre** [comment]
@simahawk We should completely get rid of p.formwidget.contenttree (and p.formwidget.autocomplete). We should use CatalogSource. See navigation.py:INavigationTile.root for example. It will automatically get the correct widget in Plone 5. Plone 4 requires extra registration, which is shown for INavigationTile.root in media.zcml (the inherited CatalogSource shown in navigation.py is required to correctly register widget in Plone 4 without danger of conflicting registration with other packages). (All file references to p.a.standardtiles.)

**Comment by thet** [comment]
Another two cent from me: recently I have done something similar for portlets, based on Plone's content full view. It uses mockup's relation widget. Maybe it can help: https://github.com/collective/collective.portlet.fullview

**Comment by datakurre** [comment]
@thet NavigationTile does exactly same, but uses inherited CatalogSource so that Plone 4 can register widget without conflicting e.g. with c.portlet.fullview's widget registration :) (@simahawk that's only issue on Plone 4, because there's no common default registration for those widget, but every add-on wants/needs to register those themselves)

**Comment by datakurre** [comment]
@simahawk @thet Yet, collective.portlet.fullview could be useful to see how properly render just some macro (preferrably content-core) from those contents' default view.

**Comment by simahawk** [comment]
@thet tnx a lot! It helped :) 

I have the tile working. One 1st main issue: rendering is broken on edit if you use '#content' as a selector. I guess it depends on some portal or JS transform. So, we must change any conflicting IDs, at least the ones we already know, to make it working always as expected.

Note: if you want to test this on an existing plone site you need to update registry conf for tiles, because it has been renamed.

**Comment by datakurre** [comment]
https://github.com/plone/plone.app.mosaic/pull/62
https://github.com/plone/plone.app.standardtiles/pull/12

**Comment by datakurre** [comment]
![nayttokuva 2015-05-05 kello 4 02 12](https://cloud.githubusercontent.com/assets/160447/7465595/8cd50426-f2db-11e4-8f10-c36efa8acb06.png)

**Comment by datakurre** [comment]
![nayttokuva 2015-05-05 kello 4 03 01](https://cloud.githubusercontent.com/assets/160447/7465599/ad2e0bfa-f2db-11e4-8e6b-ea1bb8421f38.png)

**Comment by datakurre** [comment]
What do you think about showing target content title or description? Or viewlets outside content-core? Maybe some of those should be restored as options?

**Comment by agitator** [comment]
Do we have any kind of configuration on that tile?
As default I would go with title and description for folderish types and title, description and content-core for nonfolderish.

**Comment by datakurre** [comment]
@simahawk Could you add option flags for title and description as @agiator proposed?

**Comment by datakurre** [comment]
Another, actually major, issue: What should proxy tile do with a content, which has Mosaic Layout? The current (c.p.fullview) approach fails there. Also, tiles merge is not recursive, so tile cannot return content with tiles (even if that would be possible [it could be for ESI tiles], it might be a bad feature for performance).

**Comment by djay** [comment]
Not sure how to achieve it but it would be nice if you could include an
existing tile. Then for example you could have a whole page of images or
text snippets which you reuse instead of a folder.

**Comment by datakurre** [comment]
> Not sure how to achieve it but it would be nice if you could include an
> existing tile. Then for example you could have a whole page of images or
> text snippets which you reuse instead of a folder.

@djay Definitely there should be a way to re-use (copy) existing layouts and probably also individual (configured) tiles fromlayouts. "Existing tile" could be either fully JavaScript-based feature in Mosaic Editor or a "Wizard" like transient tile.

> > Another, actually major, issue: What should proxy tile do with a content,
> > which has Mosaic Layout?

@simahawk Actually, nothing prevents us from making the composition already in tilerendering. Existing content (Proxy) tile now already detects if the target's default view is template or a browser view. If not, it should check if the view is a layout, run tile transform for the layout and then render panels from the layout.

Now the only issue would be, which panel it should render (right now we have only one panel, "content", but with site layouts we could have any amount of panels). The correct answer is probably "the same panel it is included into", but how to detect that, remains an open question.

**Comment by datakurre** [comment]
@simahawk I added title and description, but no options for them.

I also added minimal support for layouts. Currently it just renders all found panels.
There's a "minor issues" with MosaicEditor appending title from included layout into the current page title during save, though. Mosaic Editor should be fixed to not concatenate title from multiple title fields (that makes no sense).

![nayttokuva 2015-05-06 kello 9 24 13](https://cloud.githubusercontent.com/assets/160447/7488350/b288134c-f3d1-11e4-9236-8bcd95f8417d.png)

**Comment by datakurre** [comment]
We are still missing options to display title / description, but because those are not blockers, let's create new issue for them. (Yet, it may not be trivial to enforce that option to layout views – one should try to find and drop tiles with xpath searches to enforce.)

**Comment by djay** [comment]
I should explain this a little better.
It would be nice if you could select a specific tile inside another page to show in a existing content tile.
The UI would somehow to have preview each of the tiles contents since tiles don't really have names.
If we don't do this then it might seem a bit arbitrary why we can only reuse whole pages inside a single tile and not tiles inside a tile.

**Comment by datakurre** [comment]
Each apptile (even transient) have uuid and whe properly deleted, there will be event about the removal, but I'd guess those would still be even more fragile than whole page references. Currently we are completely lacking linkintegrity checking for uuid references :/
